faf-commons-data/src/main/java/com/faforever/commons/replay/semantics/Semantics.java

```diff
@@ -0,0 +1,103 @@
+package com.faforever.commons.replay.semantics;
+
+import com.faforever.commons.replay.body.Event;
+import com.faforever.commons.replay.header.Source;
+import com.faforever.commons.replay.semantics.records.ChatMessage;
+import com.faforever.commons.replay.semantics.records.TrackedEvent;
+import com.faforever.commons.replay.shared.LuaTable;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+
+public class Semantics {
+
+  public static Duration tickToDuration(int tick) {
+    return Duration.ofSeconds(tick / 10);
+  }
+
+  public static List<TrackedEvent> registerEvents(List<Source> sources, List<Event> events) {
+    final int[] tick = {-1};
+    final int[] clientId = {-1};
+
+    return events.stream().map((event) -> switch (event) {
+      case Event.Advance e -> {
+        tick[0] = tick[0] + e.ticksToAdvance();
+        yield null;
+      }
+
+      case Event.SetCommandSource e -> {
+        clientId[0] = e.playerIndex();
+        yield null;
+      }
+
+      default -> new TrackedEvent(tick[0], sources.get(clientId[0]), event);
+    }).filter(Objects::nonNull).toList();
+  }
+
+  ;
+
+  /**
+   * Retrieves all events that are chat messages
+   *
+   * @param events A list of events
+   * @return A list of events that are chat messages
+   */
+  public static List<ChatMessage> findChatMessages(List<Source> sources, List<TrackedEvent> events) {
+    return events.stream().map((trackedEvent) -> switch (trackedEvent.event()) {
+
+      // TODO: the fact that we piggy-back on the 'GiveResourcesToPlayer' callback to embed chat messages is all wrong! We should instead introduce an alternative callback with the sole purpose to send messages.
+      //  Requires refactoring in the game!
+
+      case Event.LuaSimCallback(
+        String func, LuaTable parametersLua, Event.CommandUnits commandUnits
+      ) when func.equals("GiveResourcesToPlayer") -> {
+        yield switch (parametersLua) {
+          case LuaTable.Table callbackTable -> {
```

Comment:
You don't need the internal switch. If you type parametersLua as LuaTable.Table the case will only match when that is the case
```suggestion
      case Event.LuaSimCallback(
        String func, LuaTable.Table callbackTable, Event.CommandUnits commandUnits
      ) when func.equals("GiveResourcesToPlayer") -> {
```

Reply:
And since this is only one case label we can just use an if (!( instanceof))

Reply:
I removed the internal switch in https://github.com/FAForever/faf-java-commons/pull/143/commits/7d4ef66aa3e455d10836098f9f6d0600714ba420

Reply:
I don't know how to rewrite the instance check and also keep the `when` statement, or are those two separate if statements?

Reply:
You would just change it from when to a normal && condition

Reply:
I'd rather keep the current format. It makes it consistent with the other semantic-related functions. If you want me to change it anyway then I'll do that; but I think consistency is nice.